src/main/groovy/com/google/protobuf/gradle/ToolsLocator.groovy

```diff
@@ -83,13 +86,23 @@ class ToolsLocator {
                     version: version,
                     classifier: project.osdetector.classifier,
                     ext: 'exe']
-    project.logger.info("Resolving artifact: ${notation}")
     Dependency dep = project.dependencies.add(config.name, notation)
-    File file = config.fileCollection(dep).singleFile
-    if (!file.canExecute() && !file.setExecutable(true)) {
-      throw new GradleException("Cannot set ${file} as executable")
+
+    for (GenerateProtoTask task in tasks) {
+      // register the configuration dependency as a task input
+      task.inputs.files(config)
+
+      task.doFirst {
+        if (locator.path == null) {
```

Comment:
A further optimization: you could skip the resolution if `locator != protoc && task.plugins.findByName(locator.name) == null`, thus the plugin executable won't be resolved if it's not used by any tasks that will be run.